plugins/telink-dfu/fu-telink-dfu-hid-device.c

```diff
@@ -0,0 +1,370 @@
+	FuUsbDevice *usb_dev = FU_USB_DEVICE(FU_DEVICE(self));
+	FuUsbInterface *iface = NULL;
+	g_autoptr(GPtrArray) ifaces = NULL;
+	g_autoptr(GPtrArray) endpoints = NULL;
+	FuUsbEndpoint *ep;
+	guint ep_dir;
+	guint8 ep_num;
+	guint16 maxpktsize;
+
+	ifaces = fu_usb_device_get_interfaces(usb_dev, error);
+
+	for (guint i = 0; i < ifaces->len; i++) {
+		iface = g_ptr_array_index(ifaces, i);
+		endpoints = fu_usb_interface_get_endpoints(iface);
+		g_debug("intf %u class=%u", i, fu_usb_interface_get_class(iface));
+		for (guint j = 0; j < endpoints->len; j++) {
+			ep = g_ptr_array_index(endpoints, j);
+			ep_dir = fu_usb_endpoint_get_direction(ep);
+			ep_num = fu_usb_endpoint_get_address(ep);
+			maxpktsize = fu_usb_endpoint_get_maximum_packet_size(ep);
+			g_debug("ep=0x%02x(%s), maxpktsize=%u",
+				ep_num,
+				ep_dir == FU_USB_DIRECTION_DEVICE_TO_HOST ? "in" : "out",
+				maxpktsize);
+		}
+	}
+#endif
+}
+
+static void
+fu_telink_dfu_hid_device_show_descriptors(FuTelinkDfuHidDevice *self, GError **error)
+{
+#if defined(HAVE_HIDRAW_H) && FU_TELINK_DFU_HID_DEVICE_SHOW_DESC_INFO
+	FuHidDevice *hid_dev = FU_HID_DEVICE(self);
+	g_autoptr(GPtrArray) descriptors = NULL;
+	g_autoptr(GString) desc_err = g_string_new(NULL);
+
+	descriptors = fu_hid_device_parse_descriptors(hid_dev, error);
+	if (descriptors == NULL)
+		return;
+	g_debug("descriptors=%u", descriptors->len);
+	for (guint i = 0; i < descriptors->len; i++) {
+		g_autoptr(GError) err = NULL;
+		if (desc_err->len > 0) {
+			g_string_append(desc_err, ", ");
+		}
+		g_string_append_printf(desc_err,
+				       "descriptor 0x%x: %s",
+				       i,
+				       err->message);
+	}
+#endif
+}
+
+static gboolean
+fu_telink_dfu_hid_device_setup(FuDevice *device, GError **error)
+{
+	FuTelinkDfuHidDevice *self = FU_TELINK_DFU_HID_DEVICE(device);
+
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_telink_dfu_hid_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	fu_device_set_name(device, "telink-hid-device");
+
+	fu_telink_dfu_hid_device_show_descriptors(self, error);
+	fu_telink_dfu_hid_device_show_edp_info(self, error);
+
+	/* success */
+	return TRUE;
+}
+
+static FuStructTelinkDfuHidPkt *
+fu_telink_dfu_hid_device_create_packet(FuTelinkDfuCmd cmd, const guint8 *buf, gsize bufsz, GError **error)
+{
+	FuStructTelinkDfuHidPkt *pkt = fu_struct_telink_dfu_hid_pkt_new();
+	guint8 preamble[6] = {
+		FU_TELINK_DFU_HID_DEVICE_REPORT_ID,
+		FU_TELINK_DFU_HID_DEVICE_OP_CODE,
+		20,
+		0x00,
+		(guint8)(cmd & 0x00ff),
+		(guint8)(cmd >> 8),
+	};
+	guint8 payload_len = 0;
+
+	switch (cmd) {
+	case FU_TELINK_DFU_CMD_OTA_FW_VERSION:
+		payload_len = 0;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_START:
+		payload_len = 2;
+		break;
+	case FU_TELINK_DFU_CMD_OTA_END:
+		payload_len = 6;
+		break;
+	default:
+		/* OTA data */
+		payload_len = 20;
+		break;
+	}
+	preamble[2] = payload_len;
+
+	fu_struct_telink_dfu_hid_pkt_set_preamble(pkt, preamble, sizeof(preamble), error);
```

Comment:
need to check return value here

Reply:
the returned status of setting struct members?